File: Clojure/problem045.clj
#! /usr/bin/env clojure
; Problem 45 - Project Euler
; http://projecteuler.net/index.php?section=problems&id=045

(defn triangle [n]
  (/ (* n (inc n))
     2))

(defn penta? [n]
  (let [pn (int
            (/ (+ 1 (Math/sqrt (inc (* 24 n))))
               6))
        penta (/ (* pn (dec (* 3 pn)))
                 2)]
    (= n penta)))
    
(defn hexa? [n]
  (let [hn (int
            (/ (+ 1 (Math/sqrt (inc (* 8 n))))
               4))
        hexa (* hn (dec (* 2 hn)))]
    (= n hexa)))

(loop [n 286]
      (let [tn (triangle n)]
        (if (and (penta? tn) (hexa? tn))
          (println (list n (int tn)))
          (recur (inc n)))))

File: Python/problem045.py
#! /usr/bin/env python3
"""
Problem 45 - Project Euler
http://projecteuler.net/index.php?section=problems&id=045
"""
import math


def triangle(n):
    return n * (n + 1) / 2


def ispenta(pn):
    n = (1 + math.sqrt(24 * pn + 1)) / 6
    return n.is_integer()


def ishexa(hn):
    n = (1 + math.sqrt(8 * hn + 1)) / 4
    return n.is_integer()


if __name__ == "__main__":
    n = 286
    while True:
        tn = triangle(n)
        if ispenta(tn) and ishexa(tn):
            print(n, int(tn))
            break
        n += 1
